velox/exec/trace/QueryDataWriter.cpp

```diff
@@ -45,22 +53,32 @@ void QueryDataWriter::write(const RowVectorPtr& rows) {
   batch_->append(rows);
   dataType_ = rows->type();
 
+  const auto fileSize = dataFile_->size();
   // Serialize and write out each batch.
   IOBufOutputStream out(
       *pool_, nullptr, std::max<int64_t>(64 * 1024, batch_->size()));
   batch_->flush(&out);
   batch_->clear();
   auto iobuf = out.getIOBuf();
   dataFile_->append(std::move(iobuf));
+  if (FOLLY_UNLIKELY(
+          updateAndCheckTraceLimitCB_(dataFile_->size() - fileSize))) {
+    finish();
+  }
 }
 
 void QueryDataWriter::finish() {
+  if (finished_) {
+    return;
+  }
+
   VELOX_CHECK_NOT_NULL(
       dataFile_, "The query data writer has already been finished");
   dataFile_->close();
   dataFile_.reset();
   batch_.reset();
   writeSummary();
+  finished_ = true;
```

Comment:
Add exceedTraceLimit_ flag here, and we still allow to proceed to close trace file and write summary which indicate whether has the exceeded limit? Thanks!

Reply:
The `write` method will finish early if it exceeds the limit, which will close the trace file and write the summary but without an exceeding limit indicator.

I will add a parameter named `exceedTraceLimit` in the finish method. It is false by default and set to be true explicitly when call early finish?